### ISSUE TITLE
Tracking issue for `&mut T` in const contexts (const_mut_refs)

### ISSUE THREAD

**Issue by ExpHP** [rust-lang/rust#57349]
The `const_mut_refs` feature gate allows the following in const context:

* `&mut` arguments
* mutable (re)borrows and raw pointers (`&mut` and `addr_of_mut!` expressions)
* dereferencing mutable references and raw pointers (`*` expressions)

###  Remaining work

* [x] https://github.com/rust-lang/rust/issues/79738 `static mut FOO: &'static mut i32 = &mut 42;` is problematic, because LLVM currently duplicates the inner mutable reference when you do `static mut BAR: &'static mut i32 = FOO;`. It would probably suffice to give all nested allocations in a static names relative to their original static item. Named allocations are neither deduplicated nor duplicated.
* [x] Currently the interning of the final result of the `const` relies on types more than it should, given that `transmute` is available in consts. Fixed in https://github.com/rust-lang/rust/pull/119044.
* [x] https://github.com/rust-lang/rust/issues/120450
* [ ] Figure out what to do with https://github.com/rust-lang/rust/issues/121610: that's happening in https://github.com/rust-lang/rust/pull/128543
* [ ] Adjust documentation: https://github.com/rust-lang/reference/pull/1590
* [ ] Stabilization PR: https://github.com/rust-lang/rust/pull/129195

### Open questions

* [x] can we stabilize this without figuring out the const heap story?  -> yes, no reason this has to wait for const_heap
* [ ] can we stabilize this without cranelift and codegen_gcc support? https://github.com/rust-lang/rust/issues/122399, https://github.com/rust-lang/rustc_codegen_gcc/issues/473

**Comment by Centril** [comment]
Thanks for filing this issue. :)

> * non-initialization assignments could be made (`x = y;` or `*x = y;`).

The first half of this, i.e. `x = y`, is covered by https://github.com/rust-lang/rust/issues/48821 which is being stabilized in https://github.com/rust-lang/rust/pull/57175.

**Comment by alercah** [comment]
I could be wrong, but based on previous conversation with @eddyb, this is very much in the plans. They originally wanted to even allow CTFE to mutate statics during the scope of their execution, until it was discussed out that this is undesirable because it would allow a single function call to behave differently depending on whether or not CTFE was used to precompute the function.

**Comment by eddyb** [comment]
Please cc @oli-obk and/or @RalfJung on all CTFE-related things.

In terms of the implementation, we plan to allow all pointers where the memory they point to is on the CTFE "stack" (i.e. not a static) and maybe "heap" in the future, with maybe some limitations around `const fn` itself.

But miri can already interpret more than everything you'd reasonably want to run in CTFE.

**Comment by elichai** [comment]
I think this is really useful.

i.e. pre-seeding a hash function with a context specific data:

```
pub const fn context_hash(context: &'static [u8]) -> Sha256 {
	let mut h =	Sha256::new();
	h.input(context);
	let res = h.result_reset();
	h.input(&res);
	h.input(&res);
	h
}
```

This can give a good performance boost in some applications

**Comment by tema3210** [comment]
Generally, const function must not have side effects, function to stay pure(what equals) must:
1. Do not change anything, references from outside world point to.
2. Do not use static values as they can be changed.
3. Do not change anything inside types providing interior mutability.

&mut to something can be used as usize for arithmetic, so in this case it will not cause side effect.

**Comment by cramertj** [comment]
Should this issue be blocking `const EMPTY_MAX:  &'static mut [()] = &mut [(); ::std::usize::MAX]`? Currently it does, but I think we should be able to allow this (as well as `= &mut [];` for a slice of any type).

**Comment by eddyb** [comment]
@cramertj That (`&'static mut [T; 0]`) is a silly artificial limitation AFAIK, with a FIXME in the code, we just never fixed it (it's limited to runtime `fn`s, IIRC) - feel free to submit a PR and maybe we can FCP that or w/e. This issue is about non-`'static` `&mut` in CTFE.

`&'static mut T` where `T` is a ZST is an orthogonal extension of that, which we even used to have to some extent, but we got rid of it when we RFC'd/stabilized rvalue-to-`'static` promotion.

I'm not sure we have any issue for it, maybe you can create one? It would also help to provide realistic examples for ZST expressions you want to get a `&'static mut` out of (e.g. `[(); n]` vs `[Struct {...}; n]` vs `[Foo::new(); n]` - the last one might need to use actual `layout_of` which may be more controversial).

**Comment by RalfJung** [comment]
> `[Foo::new(); n]`

No please no function calls in implicit promotion.

**Comment by eddyb** [comment]
Ah fair, I was thinking more from the fact that the value doesn't really matter if it's a ZST, but if there are preconditions (e.g. assertions) they would need to run first anyway.

**Comment by oli-obk** [comment]
I'm nominating this for @rust-lang/lang consideration for stabilization. This feature gate allows `&mut` references in arguments, return values and local variables in `const fn`. They are not permitted in any other const context, so you can't use mutable references in const items, static items, array length initializers, const generics or enum discriminant initializers. The reason is that we aren't sure how to enforce that no mutable references escape to the final value of the constant. With `const fn` there's no danger of this, because the borrow checker forbids returning mutable references to local variables. Any mutable references in the arguments that could escape to the return type must by construction be created within another const fn, as they can't even be created in any other context.

Examples:

```rust
const fn foo<'a>(x: &'a mut i32) -> &'a mut i32 {
     let mut y = 42;
     let z = &mut y;
     *z = 99;
     x
}
```

is legal, but

```rust
const FOO: () = {
     let mut y = 42;
     let z = &mut y;
     *z = 99;
};
```

is not, because we want to prevent

```rust
const FOO: &mut i32 = &mut 42;
```

and the easiest way to do this is to forbid mutable references. #71212 is tracking adding mutable references to const contexts other than `const fn` again, but that's actually not terribly important compared with all the use cases for mutable references in `const fn`.

Said uses cases being several of the issues that link to this issue (scroll up) and more concretely:

* `MaybeUninit::zeroed` (https://github.com/rust-lang/rust/issues/62061)
* `std::mem::replace`
* User functions which currently have to pass an argument and returning it again in order to mutate a value via a function call

There are no open questions known to me (beyond the use of `&mut` outside of const fn) and the implementation in the miri engine is heavily tested in the miri tool since the early days of miri.

**Comment by Havvy** [comment]
You can use mutable references in const items with just this by putting your expression into a const function and then calling the function. I even made a macro_rules macro that can do it for you ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ed5bea303f1cf0a6bff64a4d48051bd1))

```rust
macro_rules! mut_in_const_item {
    (const $id:ident: $ty:ty = $e:expr) => {
        const $id: $ty = {
            const fn __mut_in_const_item() -> $ty {
                $e
            }
            
            __mut_in_const_item()
        };
    }
}
```

Is there a reason the compiler can't do this for other `const` context? Should this question be asked in a different place?

**Comment by oli-obk** [comment]
> Should this question be asked in a different place?

yes :laughing: in https://github.com/rust-lang/rust/issues/71212

We can do an analysis that figures out that using `&mut` is fine, we just need to implement it, #71212 tracks that.

**Comment by withoutboats** [comment]
I have two questions:

1. Can you estimate the work required for #71212?
2. Are there other cases where things are allowed in const fn but not other consts right now?

**Comment by oli-obk** [comment]
> Are there other cases where things are allowed in const fn but not other consts right now?

No, there is no precedent for this. We have other differences between different const contexts though (e.g. constants cannot even mention statics, but statics can mention other statics).

>     1. Can you estimate the work required for #71212?

@ecstatic-morse is the expert on dataflow, maybe they know what needs to be done to exclude any mutable references in the final expression of a constant. We're already forbidding

```rust
const FOO: &i32 = {
    let x = 42;
    &x
};
```

so all we need to figure out is how to forbid

```rust
const FOO: &i32 = {
    let x = 2;
    &mut (1 + x)
};
```

because

```rust
const FOO: &i32 = {
    let x = 2;
    &(1 + x)
};
```

is ok.

**Comment by withoutboats** [comment]
Why is it a problem if the type of the constant contains no mutable references (as in the example you wish to forbid)?

**Comment by oli-obk** [comment]
The mutability of the thing could be hidden. Take for example a user-written wrapper that emulates mutable references: 

```rust
pub struct MutRef<T>(*mut T);

impl<T> MutRef<T> {
    pub const fn new(x: &'static mut T) -> Self {
        Self(x)
    }
    pub fn into_mut_ref(self) -> &'static mut T {
        unsafe {
            &mut *self.0
        }
    }
}
```

This would allow you to create a

```rust
const FOO: MutRef<i32> = MutRef::new(&mut 42);
```

and then call `FOO.into_mut_ref()`, because constants aren't moved out of, they are implicitly copied even if they are `!Copy`.

**Comment by oli-obk** [comment]
Unnominating, let's give mutable references outside const fn another try. I'll post an idea in #71212

**Comment by hudson-ayers** [comment]
Sometime between `nightly-2020-06-03` and `nightly-2020-10-25` the compiler changed such that code that previously worked with only a `const_fn` feature now also requires `const_mut_refs` (see https://github.com/tock/tock/pull/2175). A minimal example is here:

```rust
pub struct Foo {
    bar: Cell<Option<&'static mut u32>>,
}

impl Foo {
    pub const fn new() -> Self {
        Self {
            bar: Cell::new(None)
        }
    }
}
```

This limitation feels really artificial, because there is not actually a mutable reference being used in `const` context. I assume it is just difficult to implement the const-mutable-reference-checking in such a way that this is allowable?

**Comment by RalfJung** [comment]
> Sometime between nightly-2020-06-03 and nightly-2020-10-25 the compiler changed such that code that previously worked with only a const_fn feature now also requires const_mut_refs (see tock/tock#2175)

Yes, there is work going on towards removing the general `const_fn` feature and having more specific sub-features instead. In fact, just the `const_mut_refs` feature [should be enough](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7d3cdf980d4bbf262baae6f2c9c50a8b).

So, there is not really a new restriction; this is just the old restriction under a more specific name.

**Comment by oli-obk** [comment]
Adding to that, the limitation is indeed artificial, because we rather implemented a conservative check than accidentally allow something unsound. We have just recently made some changes that should allow us to try to tackle mutable references in constant items again.

**Comment by WaffleLapkin** [comment]
What's the current progress on this? Will mut refs are allowed in `const fn`s in any near future?

The feature was nominated for stabilization almost a year ago, so I assume it's ready for the stabilization?

**Comment by oli-obk** [comment]
We need another implementation change first (making them work in constants, too). I have most of that implemented in a local branch, I am moderately confident I'll get it done over the holidays.

**Comment by josephlr** [comment]
I think the implementation change @oli-obk is referring to is #78578, so I assume that needs to merge before stabilization can occur.

**Comment by RalfJung** [comment]
Besides that, the approach in #78578 also is highly experimental and a lot of discussion (possibly an RFC) needs to be had before we can stabilize it as-is, see https://github.com/rust-lang/rust/issues/80384 for some of the concerns.

**Comment by josephlr** [comment]
@RalfJung do we need to resolve all the issues with `const_refs_to_cell` before merging this feature (`const_mut_refs`)?

**Comment by RalfJung** [comment]
I am honestly not sure of the status here. AFAIK mutable refs inside `const fn` are okay(-ish), but to permit `&mut` in a `const`/`static` body we need to use the same analysis as `const_refs_to_cell`. This might or might not be already implemented for mutable references, and if it is then the same concerns as with `const_refs_to_cell` apply.

@oli-obk hopefully knows more than I do. :D

**Comment by oli-obk** [comment]
We allow mutable references and immutable cell references in all const contexts now. The problems mentioned for `const_refs_to_cell` apply to mutable references as follows

* `static mut FOO: &'static mut i32 = &mut 42;` is problematic, because LLVM currently duplicates the inner mutable reference when you do `static mut BAR: &'static mut i32 = FOO;`. This is not the expected behaviour, but we could just go ahead and document this new `static mut` footgun. It's not like `static mut` is really something we care about for having safely, as all uses of it are unsafe.
* Not a problem, just a thing to be aware of: We use a trivial "does this local have StorageDead" analysis, which guarantees us that none of the problematic references are in the final value of the constant, as they don't outlive the constant. A lot of things depend on this behaviour for correctness, so us also depending on it is not a new thing.
* const heap will probably allow `static FOO: &'static mut i32 = ...;` (not a `static mut`, just a regular `static`), which isn't a problem, because all accesses to `FOO` go through an immutable reference, and `&&mut X` is just like `&&X`, you don't get any new features. More problematic is `const FOO: &'static mut i32 = ...;`, but I think we need to deal with things like that *anyway* once we get heap things, as users can also just do `const FOO: MyBox<i32> = ...;`, which has all of the same problems.

So TLDR: Imo we should just allow mutable references, even if immutable cell references still need some work due to `static FOO: &'static AtomicUsize` duplicating the memory and using a different address for the value of `static BAR: &'static AtomicUsize = FOO;`

**Comment by RalfJung** [comment]
> This is not the expected behaviour, but we could just go ahead and document this new static mut footgun. It's not like static mut is really something we care about for having safely, as all uses of it are unsafe.

For unsafe primitives it is even more important than for safe ones that their behavior makes sense, since we don't have the type system doesn't help us get it right. So I very strongly object to the sentiment "it's unsafe so we don't care if it's behavior is broken".

And "broken" is the only description I can have for this kind of behavior. Furthermore, as you said yourself it also affects shared references, which is already observable with pointer equality and will be blatantly obvious with interior mutability. So, this behavior will change one day. Stabilizing this risks someone relying on the current behavior (Hyrum's Law), so we shouldn't do that.

> Not a problem, just a thing to be aware of: We use a trivial "does this local have StorageDead" analysis, which guarantees us that none of the problematic references are in the final value of the constant, as they don't outlive the constant. A lot of things depend on this behaviour for correctness, so us also depending on it is not a new thing.

I think currently it would be totally correct to just remove all these annotations (the program might just become slower). Currently these Storage\* annotations are useful for optimizations, but we have a lot of freedom in adjusting their design, and some proposals for that exist (like `StorageRefresh`). There's also a lot of them, bloating the MIR, so finding ways to have fewer of them could significantly reduce build times. We already remove them early for debug builds, maybe we can avoid emitting them entirely for debug builds?

In contrast, this const analysis makes the delicate details of when and where these annotations are placed publicly observable by stable code. That's a big commitment; after stabilization, *any* change the how we generate those annotations could easily be a breaking change.

Furthermore (but I am less worried about this second point), the analysis relies on some structural properties of these annotations (I forgot the details but IIRC there are comments explaining this in the code), and we don't currently have any check that would help us notice when these structural properties stop holding.

SO my TLDR is, we need to fix the allocation duplication thing (or reject `&mut` in `static mut`), and we better be really sure that we want to commit to keeping something very close to our current StorageLive/StorageDead system -- even when Polonious will completely change how the borrow checker works.

**Comment by RalfJung** [comment]
> More problematic is const FOO: &'static mut i32 = ...;

We just don't allow that, right? The really problematic cases I think are things like
```rust
const FOO: *const i32 = &mut 13 as *mut _ as *const _;
```
but it actually catches that.

**Comment by oli-obk** [comment]
> For unsafe primitives it is even more important than for safe ones that their behavior makes sense, since we don't have the type system doesn't help us get it right. So I very strongly object to the sentiment "it's unsafe so we don't care if it's behavior is broken".

Sorry, I meant, static mut is frequenly unsound already and should be deprecated anyway: https://github.com/rust-lang/rust/issues/53639

**Comment by oli-obk** [comment]
> I think currently it would be totally correct to just remove all these annotations (the program might just become slower)

No, borrowck would end up with unsoundness if we removed all storagedead. It must know when things are finished. Maybe it would not be unsound, but it would at least be a breaking change because suddenly borrowck things change in weird ways.

**Comment by RalfJung** [comment]
> Sorry, I meant, static mut is frequenly unsound already and should be deprecated anyway: #53639

It has been proposed for deprecation several times, but has not actually been deprecated as such, so I don't think we can treat it as such.

> No, borrowck would end up with unsoundness if we removed all storagedead. It must know when things are finished. Maybe it would not be unsound, but it would at least be a breaking change because suddenly borrowck things change in weird ways.

Hm, I see.
But the structure of borrowck will change a lot with Polonius, and it is not at all clear if the approach we currently have can be ported to the logic programming approach of Polonius.

>  If we replace them with something else that borrowck uses, we will use them, too

This is the thing that I have no confidence will actually be possible. There might not be any syntactic markers in the code with Polonius, just some implicit constraints gathered in a logic program.

**Comment by eddyb** [comment]
> But the structure of borrowck will change a lot with Polonius, and it is not at all clear if the approach we currently have can be ported to the logic programming approach of Polonius.

That's just the part that reasons about "regions" of code for which a borrow is held, isn't it? The special nature of `Drop` terminators and `StorageDead` terminators would still be preserved, where e.g. a variable is considered uninitialized after either.

> This is the thing that I have no confidence will actually be possible. There might not be any syntactic markers in the code with Polonius, just some implicit constraints gathered in a logic program.

AFAIK MIR will not change for Polonius, the logic constraints are computed from MIR within the confines of the borrowck analysis (just as today). But maybe @nikomatsakis can more explicitly confirm that `StorageDead` is here to stay.

**Comment by jplatte** [comment]
The issue description says

> The `const_mut_refs` feature gate allows all of the above in `const fn`, but not in other const items. Allowing mutable references outside of `const fn` is tracked in #71212

but that issue was closed by #78578, which AFAICT just extended the `const_mut_refs` feature gate to all const contexts? If I understood correctly I think the issue description should be updated.

**Comment by bstrie** [comment]
Since part of this feature was proposed for stabilization a few years ago now, that would seem to imply that in certain contexts `&mut`-in-`const` is ready to land, and that the remaining problematic contexts are keeping everyone on stable from being able to leverage this in the non-problematic contexts. Does that seem correct? If so, can we split this out into multiple features and stabilize the parts that people are confident in?

**Comment by clarfonthey** [comment]
Would someone be willing to compile a list of the current issues with this that are blocking stabilisation, and/or what subsets could be split out to be stabilised?

Right now it's a bit difficult to tell what the current status of this is, and the OP doesn't have the usual tracking issue task list. If someone with permission to edit it could add said list as well, that would be lovely.

**Comment by oli-obk** [comment]
> subsets could be split out to be stabilised?

We tried stabilizing a subset, but that didn't work out. There are no subsets that we can stabilize. It's also completely out of my cache what is necessary to stabilize the full feature. I'll assign the issue to me so I pick it up when I have time in the next weeks

**Comment by oli-obk** [comment]
It didn't let me go, so I read the entire impl history of immutable references with interior mutability and of mutable references. I added the remaining work item to the main post of this issue. Not sure how to approach it yet, we may have to add a new field to Allocation that contains the DefId of the static that it's a part of.

**Comment by oli-obk** [comment]
> Ah, so as usual, `static mut` ruins everyone's day.


Well, for interior mutability reular statics have the exact same problem. So by fixing it here, we also fix it for interior mutability in immutable statics

**Comment by RalfJung** [comment]
IOW: as usual, mutability ruins everyone's day. Whether it is via `static mut` or `static _: UnsafeCell` makes little difference. ;)

**Comment by chriss0612** [comment]
Is there a reason why functions pointers with mutable references in Argument or return Position need this feature?
Since fn pointers can't be called in const fn's anyways it shouldn't be a problem to allow those.

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=c967ceadcc4cfdca1b93793e1096ea8d)

**Comment by oli-obk** [comment]
That's just a bug, I'd just prefer not to add exceptions to the check and instead stabilize the feature

**Comment by AJAnderson** [comment]
Is there a work around for this currently? I'm trying to get something like the following to compile:

```
static mut GLOBAL_NAME: GlobalStruct = GlobalStruct::new()

struct GlobalStruct {
    data_field: DataStruct,
    func_pointer: Option<fn(&mut GlobalStruct)>
}

impl GlobalStruct {
    fn new() -> Self {
         data_field: Default::default(),
         func_pointer: None
    }
}
```

**Comment by oli-obk** [comment]
There is no workaround, but @pvdrz is actively working on fixing the blocking bug, and his current branch looks very promising

**Comment by Boiethios** [comment]
Is it normal that I cannot write this const function?

```rust
pub const fn generate_increment() -> &'static dyn Fn(&mut u8) {
    &|u: &mut u8| *u += 1
}
```

I get:

```
error[[E0658]](https://doc.rust-lang.org/stable/error-index.html#E0658): mutable references are not allowed in constant functions
 --> src/lib.rs:1:38
  |
1 | pub const fn generate_increment() -> &'static dyn Fn(&mut u8) {
  |                                      ^^^^^^^^^^^^^^^^^^^^^^^^
  |
```

But there is not really any mutable reference inside the function. It's inside the closure, which isn't really related to the function.

**Comment by c410-f3r** [comment]
> ```rust
> pub const fn generate_increment() -> &'static dyn Fn(&mut u8) {
>     &|u: &mut u8| *u += 1
> }
> ```

It is necessary to use a newer compiler version -> https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=26c67b4f419389df9a045e2c8f22f5ee

**Comment by hudson-ayers** [comment]
> > ```rust
> > pub const fn generate_increment() -> &'static dyn Fn(&mut u8) {
> >     &|u: &mut u8| *u += 1
> > }
> > ```
> 
> It is necessary to use a newer compiler version -> https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=26c67b4f419389df9a045e2c8f22f5ee

I think the original commenter meant "is it normal that I cannot write this function without this feature", since the function does not actually use the mutable reference in question. But I imagine that this is roughly the same problem as the one in https://github.com/rust-lang/rust/issues/57349#issuecomment-1365706366 just above, which also has not been solved yet

**Comment by Boiethios** [comment]
> > > ```rust
> > > pub const fn generate_increment() -> &'static dyn Fn(&mut u8) {
> > >     &|u: &mut u8| *u += 1
> > > }
> > > ```
> > 
> > 
> > It is necessary to use a newer compiler version -> https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=26c67b4f419389df9a045e2c8f22f5ee
> 
> I think the original commenter meant "is it normal that I cannot write this function without this feature", since the function does not actually use the mutable reference in question. But I imagine that this is roughly the same problem as the one in [#57349 (comment)](https://github.com/rust-lang/rust/issues/57349#issuecomment-1365706366) just above, which also has not been solved yet

So, it's actually a bug, thank you for the answer. I scanned quickly the previous comments but didn't find this. Sorry for the noise.

**Comment by wrenger** [comment]
Having &mut refs on const functions would greatly simplify my bitfield crate (which generates const accessor functions). 
How far are we from stabilizing this (neat 😉) feature?

**Comment by jhpratt** [comment]
By itself? No. It would also require const traits and const trait impls.

**Comment by wrenger** [comment]
@daniel-pfeiffer, if I'm correct, const traits are not stable yet. For relies on the Iterator trait.

**Comment by daniel-pfeiffer** [comment]
Thanks! In that case someone might find my workaround useful in the meantime. , ; => happened by coincidence, as they are the only separators the compiler would allow after min
```rust
// forish!{ i in 0, 9 => { ... }} is for i in 0..9 { ... }
// forish!{ i in 0; 9 => { ... }} is similar, but ... is responsible for modifying i
macro_rules! forish {
    ($name:ident in $min:expr, $max:expr => $body:block) => {
        forish!($name in $min; $max => { $body; $name += 1; })
    };
    ($name:ident in $min:expr; $max:expr => $body:block) => {
        let mut $name = $min;
        while $name < $max $body
    }
}
```

**Comment by ia0** [comment]
Is it normal for E0658 to trigger here?

```rust
pub const fn none() -> Option<&'static mut u8> {
    None
}
```

```
error[E0658]: mutable references are not allowed in constant functions
 --> src/lib.rs:1:24
  |
1 | pub const fn none() -> Option<&'static mut u8> {
  |                        ^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #57349 <https://github.com/rust-lang/rust/issues/57349> for more information
  = help: add `#![feature(const_mut_refs)]` to the crate attributes to enable
```

I would argue that there are no mutable references here. Note also that the following works (which I would argue is similar):

```rust
static FOO: Option<&'static mut u8> = None;
```

In practice the type I'm using is not `Option<&'static mut u8>` but `critical_section::Mutex<RefCell<Foo<&'static mut T>>>` where `Foo<T>` is an enum similar to `Option<T>` and `T` is a generic parameter of the `const fn`.

**Comment by griffi-gh** [comment]
Can't even handle function pointers that contain mut refs without this... (even casting them to raw pointers is not allowed)

**Comment by nayeemrmn** [comment]
I believe I'm getting a false-negative for this error:
```rust
const fn get_fn_pointer() -> fn(&mut String) {
  String::clear
}
```
```
error[E0658]: mutable references are not allowed in constant functions
   --> src/main.rs:103:30
    |
103 | const fn get_fn_pointer() -> fn(&mut String) {
    |                              ^^^^^^^^^^^^^^^
    |
    = note: see issue #57349 <https://github.com/rust-lang/rust/issues/57349> for more information
```
Is this a rustc bug?

**Comment by RalfJung** [comment]
If anything this would be a false positive, since you are getting an "error detected" message.

But this is currently intentional.

**Comment by Lokathor** [comment]
The two comments before yours (possibly more) were reporting the same problem you're seeing. It's definitely a longstanding issue, not some short and temporary bug.

**Comment by nayeemrmn** [comment]
Thanks for responding.

> But this is currently intentional.

Does this warrant a separate issue (if there isn't one) / some intermediary fix? Or does it fall under the same umbrella in some obvious way that I'm missing. The error message doesn't seem applicable at all.

**Comment by nayeemrmn** [comment]
Here's a more egregious example:
```rust
const fn some_const_fn1() {
  let _f = String::clear;
  // This works.
}

const fn some_const_fn2() {
  let _f: fn(&mut String) = String::clear;
  // error[E0658]: mutable references are not allowed in constant functions
}
```

**Comment by jhpratt** [comment]
That seems like a bug to me, as explicitly naming the type that is inferred should never cause an error.

**Comment by GoldsteinE** [comment]
> That seems like a bug to me, as explicitly naming the type that is inferred should never cause an error (barring the type itself being unstable, which is not the case here).

I think that's a different type tho? Function item type coerces into a function pointer type in the second example.

**Comment by jhpratt** [comment]
Ah, right. Yay coercion.

**Comment by griffi-gh** [comment]
> I believe I'm getting a false-negative for this error:
> 
> ```rust
> const fn get_fn_pointer() -> fn(&mut String) {
>   String::clear
> }
> ```
> 
> ```
> error[E0658]: mutable references are not allowed in constant functions
>    --> src/main.rs:103:30
>     |
> 103 | const fn get_fn_pointer() -> fn(&mut String) {
>     |                              ^^^^^^^^^^^^^^^
>     |
>     = note: see issue #57349 <https://github.com/rust-lang/rust/issues/57349> for more information
> ```
> 
> Is this a rustc bug?

no, this intentionally prevents you from handling functions that use mut arguments.  
you need to wrap it in a struct and init it in a const block

**Comment by nayeemrmn** [comment]
> you need to wrap it in a struct and init it in a const block

Thanks! My actual use case is already wrapped in a struct but init-ing it in a const block may not be an option if it references type parameters. In my case I can cherry-pick the relevant fields fortunately :-)

> no, this intentionally prevents you from handling functions that use mut arguments.

It's surprising then that `const some_fn: fn(&mut String) = ...` is allowed but `const fn get_some_fn() -> fn(&mut String) {}` isn't. Hopefully that issue can be decoupled from this wider issue for `&mut` in const fn.

**Comment by RalfJung** [comment]
[A year ago](https://github.com/rust-lang/rust/issues/57349#issuecomment-1266769993), the status on `fn(&mut  _)` was

> That's just a bug, I'd just prefer not to add exceptions to the check and instead stabilize the feature

Looks like stabilizing the feature might take... who knows how long.  So maybe we should add an exception to function pointers in the mean time.

**Comment by oli-obk** [comment]
For anyone interested what the blocker for this feature is, [here's a godbolt](https://rust.godbolt.org/#z:OYLghAFBqd5TKALEBjA9gEwKYFFMCWALugE4A0BIEAZgQDbYB2AhgLbYgDkAjF%2BTXRMiAZVQtGIHgBYBQogFUAztgAKAD24AGfgCsp5eiyahSAVyVFyKxqiIEh1ZpgDC6embZMDzgDIEmbAA5TwAjbFIQAFZyAAd0JWIHJjcPLwN4xPshf0CQtnDImJtsO2SRIhZSIlTPbx5rbFtspgqqolzgsIjo60rq2vSGy3bO/MLogEprdDNSVE4uAFIAJgBmMDAlqIAhGmwWIjnsCAwmSwB9NjMiC9JsGiVJ7YARJa0AQXeP2LNQgGoRvZUP9rkR/gAxADyUJA/1WADYwf8zAAOeFrF7wlZIm7/aQrJZrHbfX4AoEEEHInYfABKcMRyLRGKxZnOLH28IA7DtITDuW9id8uNN6NwovxvFwdOR0NxaRZwUpZvNsNi1nxyERtCLpgBraLSAB0PC0prWKy00i5qK0ZoaYq40n4bCkdvIUplcq4/CUIC0Wp10zgsEQEBQGDYsQYEUo1Ej0cYkSYBGASCI9AAnuQcAA3SnYABqBGwAHcobFmNxNXR6EQIn6IKEdeRQgEqpnq/w26xSJmoaFdKVtbx%2BJGOMIoUwsy2cKEzMAXBJ6H7RznsGxjMBJNL%2BIR7mVc9hVzLsOpSjdFpqAvXHTL6ARQqQO24cC2iKQCK7R9MaEZgEoxZlhWVZroIwhiBInAyHIwjKGomi7uQ%2BgOluIDmJYhiPn6kDTOgsQtKuAC0ABabKMCwShIERRBINgRFsAQn5EQEiQ4ER6A0ERmazPw6BHqQn44DhUCsBwIDYIQLTkLmEhmIslorHwkzTCUZSOBAziDPU5B%2BAEXQFD0DSZEkQjaRkCSmUwYzdJEwxNMO5T9DU7h1AYaktG01Q2YZdl9O05nDM5PkTDwqkqgsUiiuKkott6/wYUQIIpmmGaZv8ED4MQZDqmF/AjjoKnkAaUQBo6zrkK6pUenF3C%2Bv6ga7sGYYoLMRC/FYVCnOgUYxqQQTsIsiXJam6ZZvwknZUJBjgaI4iSDBs3wRoLYoeQpbPrEXbRVwEo1Uh3pQjcHX/JxCWKiNqVZhlbi9UmuWTPlQbTHRLA4JEEA7RVVUBp6fF1dYDUFbqxUgNIURGqiACc0MIlyUNREpUMw1DhjcGssUHQDwNFY6KwutEv21T6jWFeQIbICATFxt1d0RAN4lcMN/wpWNmYTVJZAEJgM3yJBC2yEtKgrUha0eckThMK4rlDLpUshUZcSWS0gVK1kyQK354tCF5LlpDp2utMF%2BnjIrIwDDLBvG3ktlRVq9zYNz/o7Xtf2ytwvi%2BIWACy/wAJK0v8QGlhE52WJdbMZVlJCkLl5D/Ldiah6sGqPaTIOve91BfQT1Vu969UBjjO0Y/tXrY89MkNhL0hAA) that shows how allocations in static items end up getting duplicated.

**Comment by RalfJung** [comment]
Is that the only blocker? I seem to recall the main blocker was type system issues: we have to prevent thing like `const C: &mut i32 = &mut 0`, and more importantly the raw pointer and other unsafe code versions of that?

**Comment by RalfJung** [comment]
In particular I think we should never accept code like this
```rust
pub static mut FOO: &mut [i32] = &mut [42];
```
Whenever global data is mutable, we should explicitly declare it to have some identity, e.g. via
```rust
static mut FOO_DATA: [i32; 1] = [42];
pub static FOO: &mut [i32] = &mut FOO_DATA;
```
When interning the allocations of a static, ideally everything except for the top-level allocation is always read-only; that way we avoid having to implicitly guess mutability for things the user never explicitly declared.

**Comment by oli-obk** [comment]
> In particular I think we should never accept code like this

I agree, but it's been stable since before I touched const iirc

**Comment by RalfJung** [comment]
Oh, I thought I had tested this and it got rejected.

Well we can still make sure we keep rejecting this though
```rust
pub static FOO: &mut [i32] = &mut [42];
```
and `static mut` is on its way towards deprecation anyway...

**Comment by JohnTheCoolingFan** [comment]
I have encountered E0658 which linked here in my code, but I believe I don't use any mutable references in my const functions. The only mutable reference there is in a type of a field used on a struct, a function pointer that takes a mutable reference.
Is this intended behavior or should I report this as a bug?

**Comment by Nemo157** [comment]
@JohnTheCoolingFan try testing on nightly, 1.75 has a fix for https://github.com/rust-lang/rust/issues/114994 which seems relevant.

**Comment by oli-obk** [comment]
Function pointers have been fixed. You may need to use beta for a bit or wait until the next stable release

**Comment by JohnTheCoolingFan** [comment]
@Nemo157 Yes, now the error makes sense on nightly, as it only errors on usage ofd mutable references like `Option::get_or_insert_with` (which are not usable in const contexts anyway).
Thanks, glad to know this is fixed for the next release.

**Comment by yanchith** [comment]
Hi all. Not sure this is the best place to ask, but will it be ever possible to modify static values in const contexts with `const_mut_refs`? Something like this?

```rust
#![feature(const_mut_refs)]

static mut COUNTER: usize = 0;

const fn assign_next(counter: &mut usize) -> usize {
    let index = *counter;
    *counter += 1;
    
    index
}

fn main() {
    static ZONE1: usize = assign_next(unsafe { &mut COUNTER });
    static ZONE2: usize = assign_next(unsafe { &mut COUNTER });
    
    println!("ZONE1 {ZONE1}");
    println!("ZONE2 {ZONE2}");
}
```

What I am going for is something like the __COUNTER macro in C. I'd like the compiler to give me a new and unique integer each time it sees a thing in code.

**Comment by NobodyXu** [comment]
@yanchith Don't think that's possible, const function must be pure, meaning no modification to global state or I/O.

**Comment by RalfJung** [comment]
Indeed, there are no plans to allow that and it is completely unclear how it could even be specified. For instance, we would have to define *in which order* all the constants in a crate are evaluated. This is not something we want to ever make any difference, so mutating statics in consts is out of the question.

**Comment by yanchith** [comment]
Well thanks for the replies :) I'm going to think on it some more to see if I can hack it with existing Rust before resorting to meta-programming and code generators.

> define in which order

For my usecase I really don't care about the order, just that it is synchronized so that each number is unique.

**Comment by ia0** [comment]
It might not work in your case (i.e. if you need those numbers in const eval), but you could abuse the linker to allocate those unique numbers for you, the same way `defmt` does it. Essentially, for each unique number, you create a `static mut u8` in some section that you'll eventually discard. The linker will give each of them a unique address. You can cast pointers to those `static mut u8` to `usize` (but only in non-const code). So you actually have unique constants during runtime, but the compiler doesn't know the value of those constants at compile time (only your program does, after linking). Just make sure you only ever use those pointers to cast them to `usize` and never read or write from them.

**Comment by yanchith** [comment]
That's a very interesting thought. Will give it a try :) My other thought was abusing `TypeId`, but I have no idea if they are without holes.

**Comment by CryZe** [comment]
There's a crate for that: https://github.com/dtolnay/linkme

**Comment by ia0** [comment]
`linkme` solves a slightly different and more general problem but it could be used to generate unique IDs with a cost in read-only memory footprint which you don't get with a custom solution like `defmt` (a pure zero-cost abstraction to get unique IDs) because `linkme` provides you access to the content of the distributed slice which you don't need for unique IDs.

```rust
#[distributed_slice]
static IDS: [u8];

// Each call site of this macro provides a unique usize (they will be packed but won't start at zero).
macro_rules! genid {
    () => {{
        #[distributed_slice(IDS)]
        static _ID: u8 = 0; // we don't care about this value but linkme doesn't know
        &_ID as *const u8 as usize
    }};
}
```

That said, we're probably diverging too much from the original issue of `&mut` in const contexts.

**Comment by RalfJung** [comment]
Please move this discussion elsewhere, it is off-topic for this tracking issue.

**Comment by tarcieri** [comment]
Now that https://github.com/rust-lang/rust/pull/116564 has landed, what are the next steps for `const_mut_refs`?

**Comment by oli-obk** [comment]
That PR has no effect on this feature or its blockers, it just removes a blocker for the blocker 😆 I have a branch locally that I'm rebasing right now. That should finally fix https://github.com/rust-lang/rust/issues/79738 for good

**Comment by RalfJung** [comment]
[Over here](https://github.com/rust-lang/rust/issues/120450#issuecomment-1948141441) an interesting question came up -- should this be allowed? It is currently rejected
```rust
#![feature(const_mut_refs)]
use std::sync::Mutex;

static BAR: Mutex<&mut [u8]> = unsafe {
  static mut FOO: [u8; 1] = [42]; // a private static that we are sure nobody else will reference
  Mutex::new(&mut FOO)
};
```

We don't want mutable references in the final value of statics/consts... but if they point to already existing things that are legitimately mutable, then what's the issue? (Of course this is unsafe due to aliasing concerns.)

**Comment by oli-obk** [comment]
All the blockers have been resolved. So we'll need to write up a stabilization report explaining the current status plus explanations of some of the weirder cases.

**Comment by jhpratt** [comment]
Thank you @oli-obk, @RalfJung, and everyone else who has put in countless time and effort on this!

**Comment by RalfJung** [comment]
We do have a blocker still, unfortunately: the check in the new interner is still on probation due to https://github.com/rust-lang/rust/issues/121610. The mutability check in the interner is currently a lint. My entire motivation for that interner was to make that check a hard error and I feel uneasy stabilizing const_mut_refs until we can make it a hard error.

**Comment by RalfJung** [comment]
The tracking issue lists this was part of this feature

> non-initialization assignments could be made (x = y; or *x = y;).

But in fact such assignments are already stable... looks like that feature gate was split up at some point.

**Comment by tarcieri** [comment]
@RalfJung does #128543 resolve #121610, and if so, is it the only blocker for proceeding with a stabilization report for `const_mut_refs`?

**Comment by RalfJung** [comment]
Ah, I should have given an update here: with https://github.com/rust-lang/rust/pull/129508 having landed, I no longer think https://github.com/rust-lang/rust/issues/121610 is a blocker for stabilization. The stabilization PR for this (and const_refs_to_cell) is up at https://github.com/rust-lang/rust/pull/129195. :)

**Comment by clarfonthey** [comment]
Is there a central place for all of the library APIs that are blocked by this feature being stabilised (so we can look into stabilising them), or is that something that still needs to be put together?

**Comment by RalfJung** [comment]
I don't think there is. One can probably start by looking at the back references of the const_mut_refs tracking issue.

**Comment by usbalbin** [comment]
There is this thing https://rust-lang.github.io/const-eval/skill_tree.html 

However i am not sure if it is up to date

**Comment by workingjubilee** [comment]
const fn API currently awaiting checkboxes:
- const `str::as_mut`: https://github.com/rust-lang/rust/issues/130086#issuecomment-2336408562
- const `slice::split_at_mut`: https://github.com/rust-lang/rust/issues/101804#issuecomment-2334841347
- const `slice::{first,last}_mut`: https://github.com/rust-lang/rust/issues/83570#issuecomment-2334847112
- const `mem::replace`: https://github.com/rust-lang/rust/issues/83164#issuecomment-2338660862
- const `str::from_utf8_unchecked_mut`: https://github.com/rust-lang/rust/issues/91005#issuecomment-2359820672
- const `UnsafeCell::get_mut` https://github.com/rust-lang/rust/issues/88836#issuecomment-2359817772

**Comment by RalfJung** [comment]
There's more:
- https://github.com/rust-lang/rust/issues/86302
- https://github.com/rust-lang/rust/issues/111774
- https://github.com/rust-lang/rust/issues/91822

**Comment by bjoernager** [comment]
I would also like to add `char::encode_utf8`: #130512

**Comment by RalfJung** [comment]
This has *just* been implemented. Let's maybe let it bake on nightly for a bit?

**Comment by onestacked** [comment]
#67456 needs an FCP started for Stabilization PR #130403

**Comment by Rudxain** [comment]
Is there a discussion thread for this? I've searched multiple RFCs, but I can't find a single one mentioning iterators. If I'm not mistaken, this feature is a **big** step forward to support `const` iteration. We already have `const` loops, so I can feel `core::iter` is soon becoming `const` too!

**Comment by Lokathor** [comment]
requiring const trait impls means that normal iterators are most likely rather far out.

**Comment by RalfJung** [comment]
Iterators are still blocked on being able to call trait functions in const, which is unfortunately quite tricky. See:
- https://github.com/rust-lang/rust/issues/67792
- https://github.com/rust-lang/rust/issues/87575
